provider/ecs.go

```diff
@@ -0,0 +1,478 @@
+	"github.com/aws/aws-sdk-go/aws/defaults"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/cenk/backoff"
+	"github.com/containous/traefik/job"
+	"github.com/containous/traefik/log"
+	"github.com/containous/traefik/safe"
+	"github.com/containous/traefik/types"
+)
+
+var _ Provider = (*ECS)(nil)
+
+// ECS holds configurations of the ECS provider.
+type ECS struct {
+	BaseProvider `mapstructure:",squash"`
+
+	Domain           string `description:"Default domain used"`
+	ExposedByDefault bool   `description:"Expose containers by default"`
+	RefreshSeconds   int    `description:"Polling interval (in seconds)"`
+
+	// ECS lookup parameters
+	Cluster         string `description:"ECS Cluster Name"`
+	Region          string `description:"The AWS region to use for requests"`
+	AccessKeyID     string `description:"The AWS credentials access key to use for making requests"`
+	SecretAccessKey string `description:"The AWS credentials access key to use for making requests"`
+}
+
+type ecsInstance struct {
+	Name                string
+	ID                  string
+	service             *ecs.Service
+	task                *ecs.Task
+	taskDefinition      *ecs.TaskDefinition
+	container           *ecs.Container
+	containerDefinition *ecs.ContainerDefinition
+	machine             *ec2.Instance
+}
+
+type awsClient struct {
+	ecs *ecs.ECS
+	ec2 *ec2.EC2
+}
+
+func (provider *ECS) createClient() (*awsClient, error) {
+	sess := session.New()
+	ec2meta := ec2metadata.New(sess)
+	if provider.Region == "" {
+		log.Printf("No EC2 region provided, querying instance metadata endpoint...")
+		identity, err := ec2meta.GetInstanceIdentityDocument()
+		if err != nil {
+			return nil, err
+		}
+		provider.Region = identity.Region
+	}
+
+	cfg := &aws.Config{
+		Region: &provider.Region,
+		Credentials: credentials.NewChainCredentials(
+			[]credentials.Provider{
+				&credentials.StaticProvider{
+					Value: credentials.Value{
+						AccessKeyID:     provider.AccessKeyID,
+						SecretAccessKey: provider.SecretAccessKey,
+					},
+				},
+				&credentials.EnvProvider{},
+				&credentials.SharedCredentialsProvider{},
+				defaults.RemoteCredProvider(*(defaults.Config()), defaults.Handlers()),
+			}),
+	}
+
+	return &awsClient{
+		ecs.New(sess, cfg),
+		ec2.New(sess, cfg),
+	}, nil
+}
+
+// Provide allows the provider to provide configurations to traefik
+// using the given configuration channel.
+func (provider *ECS) Provide(configurationChan chan<- types.ConfigMessage, pool *safe.Pool, constraints types.Constraints) error {
+
+	provider.Constraints = append(provider.Constraints, constraints...)
+
+	pool.Go(func(stop chan bool) {
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			select {
+			case <-stop:
+				cancel()
+			}
+		}()
+
+		operation := func() error {
+			select {
```

Comment:
I agree with your comment that the https://github.com/containous/traefik/pull/1088/files#diff-20ed89f06f1c428e51205096690cb934R149 also handles the case, but only if we're waiting for the refresh timer. If we had been in the middle of an AWS http request, that is when the select was useful.